haskell-src/Concordium/Types.hs

```diff
@@ -910,5 +916,36 @@ instance S.Serialize GenesisIndex where
   put (GenesisIndex gi) = S.putWord32be gi
   get = GenesisIndex <$> S.getWord32be
 
+-- |Equivalence class of account addresses. In protocol versions 1 and 2
+-- addresses are in 1-1 correspondence with accounts. In protocol version 3 only
+-- the first 29 bytes of the address uniquely identify an account. This type
+-- wrapper is used to wrap account addresses and add different equality and
+-- hashable instances so that we can identify transactions coming from different
+-- addresses but the same account.
+--
+-- For backwards compatibility we retain the equality and hashable instances for
+-- account addresses as they were since account addresses are compared in a few
+-- places in the scheduler.
+newtype AccountAddressEq = AccountAddressEq {
+  aaeAddress :: AccountAddress
+  }
+    deriving (Show)
+
+{-# INLINE accountAddressEmbed #-}
+-- |Embed an account address into its equivalence class.
+accountAddressEmbed :: AccountAddress -> AccountAddressEq
+accountAddressEmbed = AccountAddressEq
+
+instance Eq AccountAddressEq where
+  -- compare the first 29 bytes of the address
+  AccountAddressEq (AccountAddress a1) == AccountAddressEq (AccountAddress a2) = FBS.unsafeCompareFixedByteStrings 0 29 a1 a2 == EQ
```

Comment:
It would be good to have a constant `accountAddressPrefixLength = 29` (or something to that effect) to avoid using 29 as a magic number.

Reply:
Added.